### ISSUE TITLE
SDK 7.0.200 creates false positive CS0165, Use of unassigned local variable

### ISSUE THREAD

**Issue by pieter-venter** [dotnet/roslyn#66987]
**Version Used**: 
SDK 7.0.200

**Steps to Reproduce**:

1. Create a new console application with this in Program.cs

``` c#
class Program
{
    static async Task Main(string[] args)
    {
        DateTime timestamp;

        async Task DoSomething()
        {
            timestamp = DateTime.UtcNow;
            await Task.Delay(1);
        }

        await DoSomething();
        Console.WriteLine(timestamp);
    }
}
```

2. Run `dotnet build`
3. Observe the error:

```
MSBuild version 17.5.0-preview-23061-01+040e2a90e for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
/home/myuser/cs0165/Program.cs(14,27): error CS0165: Use of unassigned local variable 'timestamp' [/home/myuser/cs0165/cs0165.csproj]
```

**Diagnostic Id**:

cs0165

**Compare to previous version**
I compared using a docker image of the previous version of the SDK.
``` shell
docker run -it mcr.microsoft.com/dotnet/sdk:7.0.103-jammy
```
Once inside, confirm version:
```
root@5328a15698be:/# dotnet --version
7.0.103
```
Then I created the same console app.
```
dotnet new console -n cs0165
```
and replace the contents of Program.cs with the code above.

```
dotnet build
MSBuild version 17.4.1+fedecea9d for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  cs0165 -> /cs0165/bin/Debug/net7.0/cs0165.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.12
```
Also confirmed that it runs

```
root@5328a15698be:/cs0165# dotnet run
02/21/2023 21:52:21
```

**Comment by pieter-venter** [comment]
Leaving this here for posterity's sake:

* Duplicate of https://github.com/dotnet/roslyn/issues/43697
Some more explanations here:
* https://github.com/dotnet/roslyn/issues/43697#issuecomment-1267389458
* https://github.com/dotnet/roslyn/pull/64482

The language rules are [defined here](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/variables.md#94433-rules-for-variables-in-local-functions). I think it could be clarified but I'm not a language/rules designer :)
